### PR TITLE
BaseAPIClient: add childSpanId parameter to log output

### PR DESCRIPTION
**Pull request by risicle**
This required a bit of tidying & refactoring to be doable in a clear, extensible and robust way.

Also add tests for the presence/absence of this childSpanId parameter, incidentally the first tests of logging behaviour for this library.

### CODE REVIEW

File: dmapiclient/base.py [hunk 1]
```diff
@@ -120,11 +114,41 @@ def _request(self, method, url, data=None, params=None):
...
+        child_span_id = next(
+            (
+                ci_headers[header_name] for header_name in (current_app.config.get("DM_SPAN_ID_HEADERS") or ())
+                if ci_headers.get(header_name)
+            ),
+            None,
+        ) if has_request_context() else None
+
+        common_log_extra = {
+            **({"childSpanId": child_span_id} if child_span_id is not None else {}),
```
Comment (katstevens) [inline dmapiclient/base.py]: This line is failing the Python 2 flake8 check - I think we can safely take the 2.7 out of the travis.yml for this repo now?

File: dmapiclient/base.py [hunk 2]
```diff
@@ -120,11 +114,41 @@ def _request(self, method, url, data=None, params=None):
...
+        del headers
+
+        # determine our final outgoing span id - find the first of DM_SPAN_ID_HEADERS which is set to something truthy
+        child_span_id = next(
+            (
+                ci_headers[header_name] for header_name in (current_app.config.get("DM_SPAN_ID_HEADERS") or ())
+                if ci_headers.get(header_name)
+            ),
+            None,
+        ) if has_request_context() else None
```
Comment (katstevens) [inline dmapiclient/base.py]: Is it worth bothering testing the scenario where `has_request_context is None` ?  (I don't think the scenarios below cover it...)

File: dmapiclient/base.py (hunk 1 above)
Comment (katstevens) [reply dmapiclient/base.py]: Though officially removing Python 2 support would probably mean a major bump. Perhaps for another time.

**Comment by risicle** [comment]
On Thursday, 21 June 2018 08:54:30 BST Kat wrote:

> 
> Is it worth bothering testing the scenario where `has_request_context is
> None` ?  (I don't think the scenarios below cover it...)

Probably yes - will squeeze one in.

File: dmapiclient/base.py (hunk 1 above)
Comment (risicle) [reply dmapiclient/base.py]: People suggested I just major-bump anyway. Done now.

**Pull request merged**